WPMedia/muckrock
# NE-319: Chunked Uploads

## Description

This re-enables chunked uploads, after they were disabled during the refactor in #41. In that refactor, I had moved away from the [S3-specific Fine Uploader build](https://docs.fineuploader.com/quickstart/02-setting_options-s3.html) because it was a little too prescriptive, and opted instead for the generic build where we could sign requests as we wanted to.

However, it turns out that the S3 build had some additional chunking-related features which were handy — so I decided to switch back to it during this PR. I still needed to make a few tweaks to the library (to skip some of the more opinionated bits) — and since it's fully deprecated/unsupported, it felt easier to copy it into the `vendor/` directory and modify there, rather than forking / importing via NPM. (If folks feel strongly though, I'm happy to set up a fork that we can load by listing the repository in the `package.json` file.)

Now, we've reached a sort of "halfway" point: our own JS manages the request-signing and lifecycle requests, while Fine Uploader handles the actual upload requests for both simple and chunked files. I recognize that it feels a little bit hacky, and I'd love suggestions to make it better — and I tried to leave as many comments as possible to highlight exactly what's going on.

## Test steps

Try out some chunked (> 5mb) and un-chunked (< 5mb) uploads! Be sure to test uploads on the following places:
- [A new request](http://dev.muckrock.com/foi/create/)
- An existing request via the `Follow Up` button
- An existing request via the `Settings > Attach Files` menu (available to superusers)

Also try deleting files, and reloading the page to make sure the session is persisted.
